certora/specs/Liquidation.spec

```diff
@@ -0,0 +1,170 @@
+methods {
+    function isRecognizedCollateralExt(address collateral) external returns (bool) envfree;
+    // unresolved calls that havoc all contracts
+    function _.isHookTarget() external => NONDET;
+    function _.invokeHookTarget(address caller) internal => NONDET;
+    function _.tryBalanceTrackerHook(address account, uint256 newAccountBalance, bool forfeitRecentReward) internal => NONDET;
+    function _.balanceTrackerHook(address account, uint256 newAccountBalance, bool forfeitRecentReward) external => NONDET;
+    function _.emitTransfer(address from, address to, uint256 value) external => NONDET;
+    function EVCHarness.disableController(address account) external => NONDET;
+    function _.computeInterestRate(address vault, uint256 cash, uint256 borrows) external => NONDET;
+    function _.onFlashLoan(bytes data) external => NONDET;
+    function _.safeTransferFrom(address token, address from, address to, uint256 value, address permit2) internal => NONDET;
+    function _.enforceCollateralTransfer(address collateral, uint256 amount,
+        address from, address receiver) internal =>  NONDET;
+
+
+    function EthereumVaultConnector.checkAccountStatusInternal(address account) internal returns (bool, bytes memory) with (env e) => 
+        CVLCheckAccountStatusInternal(e, account);
+    function EthereumVaultConnector.checkVaultStatusInternal(address vault) internal returns (bool, bytes memory) with(env e) =>
+        CVLCheckVaultStatusInternal(e);
+
+    function _.EVCRequireStatusChecks(address account) internal => NONDET;
+}
+
+
+// This returns an arbitrary account status of the prover's choosing. It is 
+// similar to NONDETing checkAccountStatus internal and is a worakround
+// for the tool not supporting NONDET for byte return values.
+persistent ghost bool accountStatusGhost;
+function CVLCheckAccountStatusInternalBool(env e, address account) returns bool {
+    return accountStatusGhost;
+}
+
+function CVLCheckAccountStatusInternal(env e, address account) returns (bool, bytes) {
+    return (CVLCheckAccountStatusInternalBool(e, account), 
+        checkAccountMagicValueMemory(e));
+}
+
+// This is using a similar pattern as CVLCheckAcountStatusInternal
+persistent ghost bool vaultStatusBool;
+function CVLCheckVaultStatusInternalBool(env e) returns bool {
+    return vaultStatusBool;
+}
+
+function CVLCheckVaultStatusInternal(env e) returns (bool, bytes) {
+    return (CVLCheckVaultStatusInternalBool(e),
+        checkVaultMagicValueMemory(e));
+}
+
+// passing
+// If violator healthy, checkLiquidation returns maxRepay and maxYield as 0.
+rule checkLiquidation_healthy() {
+    env e;
+    address liquidator;
+    address violator; 
+    address collateral;
+    uint256 maxRepay;
+    uint256 maxYield;
+
+    require oracleAddress != 0;
+
+    uint256 liquidityCollateralValue;
+    uint256 liquidityLiabilityValue;
+    (liquidityCollateralValue, liquidityLiabilityValue) = 
+        calculateLiquidityExternal(e, violator);
```

Comment:
First I just want to clarify that for the "holy grail invariant" (In HealthStatusInvariant.spec and LiquidateHealthStatus.spec), we are actually using checkLiquidity not calculateLiquidity. This is just an entirely separate rule from the holy grail. (Perhaps this was not your point, but I just wanted to be sure!)

I do also think this could be a good rule to prove! Unfortunately, this verification project has ended so I can't start working on new rules for the project. However, I did discuss with Maya about whether this rule and another one you just suggested could be explored with her team.

Reply:
I added this rule and it is passing now: https://prover.certora.com/output/65266/ed9699a14a114c0dbad76526a55ad493/?anonymousKey=f1f0a74c2c72ede7ce77f50fbf66541e8c4f03d7

It's called `checkLiquidation_healthy_reverts` in the code.